ReactiveObjC/RACSignal+Operations.m

```diff
@@ -852,22 +853,70 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
 
 - (RACSignal *)switchToLatest {
   return [[RACSignal createSignal:^(id<RACSubscriber> subscriber) {
-    RACMulticastConnection *connection = [self publish];
+    __block BOOL outerCompleted = NO;
+    __block BOOL innerCompleted = NO;
+    __block NSUInteger currentInnerIndex = 0;
+
+    RACSerialDisposable *innerDisposable = [[RACSerialDisposable alloc] init];
+    __block os_unfair_lock lock = OS_UNFAIR_LOCK_INIT;
+
+    RACDisposable *outerDisposable = [self subscribeNext:^(RACSignal *signal) {
+      NSCAssert(!signal || [signal isKindOfClass:RACSignal.class], @"-switchToLatest expects signals as values, instead we got: %@", signal);
+      signal = signal ?: [RACSignal empty];
+
+      // Ensure no further data is sent from the previous inner signal.
+      [innerDisposable.disposable dispose];
+
+      os_unfair_lock_lock(&lock);
+      NSUInteger innerIndex = ++currentInnerIndex;
+      // In most cases this should be set back to NO, but in case where a new inner signal has
```

Comment:
Not sure I get the scenario you're trying to protect yourself from. What I think you imagine is
inner completes
inner sets completion flags
outer sends next
outer completes
outer sets completion flags (before handling the next)
outer starts handling next
avoid subscribing to it since outer is already on the track to send completion

If the next is handled before the completion, then the new signal will be subscribed to and the outer won't send completed on completion.

So basically it seels like part of the 

> // Avoid subscribing to a new inner signal which may send more values if we're already going
> // to send complete via the inner or the outer signal.

portion rather than a fear of missing a subscription. So am I missing another scenario this protects against?

Reply:
Yes, this is the same scenario. I removed the docs here since the variable name is now way better.